### PR TITLE
systemd: Replace OS updates health card entry with a page notification

### LINKED ISSUE #13400: Overview page is stuck at "Checking for updates..."
Cockpit version: 210
OS: Fedora 31 OSTree
Page: Overview

If cockpit-packagekit is *not* installed, then nothing actually responds to the the "Checking for package updates..." and the spinner just keeps on going. This is with or without cockpit-ostree.

### PR DESCRIPTION
**Pull request by martinpitt**
The visual UX should be exactly the same as now, this also does not change strings. But this fixes #13400 and simplifies the code.

### CODE REVIEW

**Comment by martinpitt** [comment]
This will break some tests due to removing the specific IDs, thus only triggering one for now.

**Comment by KKoukiou** [comment]
Just tried it locally - didn't check the code yet - but it's a bit off. 
Screen Shot 2020-01-15 at 17 14 24

**Comment by martinpitt** [comment]
@KKoukiou oh, thanks: That means that if we have several notifications, they are not formatted as a list. Indeed it is rather weird that `page_status_notifications` is already a `<li>` instead of an `<ul>` or just a div or so..

**Comment by martinpitt** [comment]
Fixed now, hopefully the tests as well.

health0

health1

**Comment by martinpitt** [comment]
There were still regressions for [rhel-8-2](https://209.132.184.41:8493/logs/pull-13401-20200115-225757-13408aa7-cockpit-project-cockpit--rhel-8-2/log.html) and [8-2-distropkg](https://209.132.184.41:8493/logs/pull-13401-20200115-225757-13408aa7-cockpit-project-cockpit--rhel-8-2-distropkg/log.html). All fixed now.

Once more with feeling!

**Comment by martinpitt** [comment]
@mvollmer: I'd appreciate your opinion here. I'm not sure I grasp what you had in mind with type: null notifications i. e. how custom/secret they are supposed to be, and which ones we do and don't want to show on Health

We have three cases, AFAICS:

  1. "standard" ones like info or warning, which appear in Health and menu, which require attention
  2. transient ones like "in progress" which appear in Health, but not in the menu, and don't require attention
  3. Custom/type null which don't appear anywhere

The "type progress" could be replaced with a "type null" and a custom icon. That should be cleaned up for sure as it's redundant for now. But this PR would restrict (3) a bit, as it conditionalizes the appearance in the health card on "type exists or details exists". (3) would still be possible when not giving a "details" field, but using anything else.

So this may need some conceptual cleanup.

File: pkg/lib/notifications.js [hunk 1]
```diff
@@ -55,7 +55,10 @@ fields:
  An arbitrary value.  The "System" overview page might monitor a
  couple of pages for their status and it will use 'details' to display
  a richer version of the status than possible with just type and
- title.
+ title. The recognized properties are:
+
+   * icon: custom icon name (defaults to standard icon corresponding to type)
+   * link: custom link target (defaults to page name)
```
Comment (mvollmer) [inline pkg/lib/notifications.js]: What about making it so that `link: false` means "not a link"?

File: pkg/systemd/page-status.jsx [hunk 2]
```diff
@@ -47,20 +49,32 @@ export class PageStatusNotifications extends React.Component {
...
+        return ["system/services", "updates"].map(page => {
+            const status = page_status.get(page);
+            if (status && (status.type || status.details) && status.title) {
+                this.props.toggle_label && this.props.toggle_label(true);
+                let target = status.details && status.details.link;
+                if (!target)
+                    target = page;
+                const jump = () => cockpit.jump("/" + target);
+                let action;
+                if (status.type === 'progress')
```
Comment (mvollmer) [inline pkg/systemd/page-status.jsx]: Let's not tie "progress" and "not a link" together.  Let's use `link: false` for the latter.

File: pkg/systemd/page-status.jsx [hunk 3]
```diff
@@ -47,20 +49,32 @@ export class PageStatusNotifications extends React.Component {
...
-                </>);
-        } else {
-            this.props.toggle_label && this.props.toggle_label(false);
-            return null;
-        }
+        // Explicit whitelist for now, until we can get a dynamic list
+        return ["system/services", "updates"].map(page => {
+            const status = page_status.get(page);
+            if (status && (status.type || status.details) && status.title) {
+                this.props.toggle_label && this.props.toggle_label(true);
```
Comment (mvollmer) [inline pkg/systemd/page-status.jsx]: `toggle_label` is not used and doesn't make sense anymore, I'd say.  Let's remove it.

File: pkg/lib/notifications.js [hunk 4]
```diff
@@ -36,8 +36,8 @@ fields:
 
  - type (string, optional)
 
- If specified, one of "info", "warning", "error".  The shell will put
- an appropriate icon next to the navigation entry for this page, for
+ If specified, one of "progress", "info", "warning", "error".  The shell
```
[the remark this answers is not in the document]
Comment (martinpitt) [reply pkg/lib/notifications.js]: In retrospect this makes sense -- I just looked at that problem before adding custom icon support. I removed this commit.

File: pkg/systemd/page-status.jsx (hunk 2 above)
Comment (martinpitt) [reply pkg/systemd/page-status.jsx]: Yep, part of the previous comment.

File: pkg/lib/notifications.js (hunk 1 above)
Comment (martinpitt) [reply pkg/lib/notifications.js]: Sounds good! Done.

File: pkg/systemd/page-status.jsx (hunk 3 above)
Comment (martinpitt) [reply pkg/systemd/page-status.jsx]: Ack, done in a separate commit.

**Pull request merged**